subsys/debug/coredump/coredump_backend_nrf_flash_partition.c

```diff
@@ -0,0 +1,347 @@
+static size_t write_buf_pos;		    /* # of dump data bytes buffered in the write buffer */
+static size_t write_pos;		    /* # of dump data bytes already written to flash */
+static uint16_t dump_crc;		    /* CRC16 of already written or buffered bytes */
+
+static inline const struct header *get_stored_header(void)
+{
+	return (const struct header *)(PARTITION_ADDR);
+}
+
+static inline const uint8_t *get_stored_dump(const struct header *header)
+{
+	return (const uint8_t *)(PARTITION_ADDR) + header->offset;
+}
+
+static inline uint16_t calc_header_crc(const struct header *header)
+{
+	return crc16_ccitt(0xffff, (const uint8_t *)header, offsetof(struct header, header_crc));
+}
+
+static inline uint16_t calc_dump_crc(const struct header *header)
+{
+	return crc16_ccitt(0xffff, get_stored_dump(header), header->size);
+}
+
+static inline bool validate_header(const struct header *header)
+{
+	return (memcmp(header->magic, MAGIC, sizeof(MAGIC)) == 0) &&
+	       (calc_header_crc(header) == header->header_crc);
+}
+
+static inline bool validate_dump(const struct header *header)
+{
+	return calc_dump_crc(header) == header->dump_crc;
+}
+
+static inline bool is_within_partition(uint32_t offset, size_t size)
+{
+	return (offset < PARTITION_SIZE) && (size <= PARTITION_SIZE - offset);
+}
+
+static void write(uint32_t offset, const uint8_t *data, size_t size)
+{
+	if (write_error != 0) {
+		return;
+	}
+
+	if (!is_within_partition(offset, size)) {
+		write_error = -ENOMEM;
+		return;
+	}
+
+#ifdef CONFIG_NRFX_NVMC
+	for (uint32_t i = 0; i < size; i += sizeof(uint32_t)) {
+		nrfx_nvmc_word_write(PARTITION_OFFSET + offset + i, UNALIGNED_GET(&data[i]));
+	}
+	while (!nrfx_nvmc_write_done_check()) {
+	}
+#elif defined(CONFIG_NRFX_RRAMC)
+	nrf_rramc_config_t config = {.mode_write = true, .write_buff_size = 0};
+
+	nrf_rramc_config_set(NRF_RRAMC, &config);
+
+	memcpy((void *)(PARTITION_ADDR + offset), data, size);
+	barrier_dmem_fence_full();
+
+	config.mode_write = false;
+	nrf_rramc_config_set(NRF_RRAMC, &config);
+#endif
+}
+
+static void erase(uint32_t offset, size_t size)
+{
+	if (write_error != 0) {
+		return;
+	}
+
+	if (!is_within_partition(offset, size)) {
+		write_error = -ENOMEM;
+		return;
+	}
+
+#ifdef CONFIG_NRFX_NVMC
+	for (uint32_t end = offset + size; offset < end; offset += FLASH_ERASE_SIZE) {
+		(void)nrfx_nvmc_page_erase(PARTITION_OFFSET + offset);
+	}
+#elif defined(CONFIG_NRFX_RRAMC)
+	nrf_rramc_config_t config = {.mode_write = true, .write_buff_size = 0};
+
+	nrf_rramc_config_set(NRF_RRAMC, &config);
+
+	memset((void *)(PARTITION_ADDR + offset), 0xff, size);
+	barrier_dmem_fence_full();
+
+	config.mode_write = false;
+	nrf_rramc_config_set(NRF_RRAMC, &config);
+#endif
+}
+
+static int copy_stored_dump(off_t offset, uint8_t *buffer, size_t size)
+{
+	const struct header *header = get_stored_header();
+
+	if (!validate_header(header)) {
+		return 0;
+	}
+
+	if (offset >= header->size) {
+		return 0;
+	}
```

Comment:
Since these are directly passed to `coredump_nrf_flash_backend_cmd` and that thing passes result to caller, shouldn't return values be error codes indicating failure of validation and copy beyond size?